packages/workbench-screens/src/spaces-screen/NewSpacePopup.tsx

```diff
@@ -0,0 +1,146 @@
+import { NotificationEvent, NotificationType } from "@kiegroup-ts-generated/uberfire-api";
+
+interface Props {
+  onClose: () => void;
+  organizationalUnitService: OrganizationalUnitService;
+  authenticationService: AuthenticationService;
+}
+
+interface State {
+  name: string;
+  errorMessages: string[];
+  displayedErrorMessages: string[];
+}
+
+export class NewSpacePopup extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = { name: "", errorMessages: [], displayedErrorMessages: [] };
+  }
+
+  private async add() {
+    const user = await this.props.authenticationService.getUser({});
+
+    const newSpace = {
+      name: this.state.name,
+      owner: (user as UserImpl).name!,
+      defaultGroupId: `com.${this.state.name.toLowerCase()}`
+    };
+
+    const emptyName = Promise.resolve().then(() => {
+      if (!newSpace.name || newSpace.name.trim() === "") {
+        this.addErrorMessage(AppFormer.translate("EmptyFieldValidation", ["Name"]));
+        return Promise.reject();
+      } else {
+        return Promise.resolve();
+      }
+    });
+
+    const duplicatedName = Promise.resolve()
+      .then(() => this.props.organizationalUnitService.getOrganizationalUnit({ name: newSpace.name }))
+      .then(space => {
+        if (space) {
+          this.addErrorMessage(AppFormer.translate("DuplicatedOrganizationalUnitValidation", ["Space"]));
+          return Promise.reject();
+        } else {
+          return Promise.resolve();
+        }
+      });
+
+    const validGroupId = Promise.resolve()
+      .then(() => this.props.organizationalUnitService.isValidGroupId({ proposedGroupId: newSpace.defaultGroupId }))
+      .then(valid => {
+        if (!valid) {
+          this.addErrorMessage(AppFormer.translate("InvalidSpaceName", []));
+          return Promise.reject();
+        } else {
+          return Promise.resolve();
+        }
+      });
+
+    this.setState({ errorMessages: [] }, () =>
+      Promise.resolve()
+        .then(() => Promise.all([emptyName, duplicatedName, validGroupId]))
+        .then(() => this.props.organizationalUnitService.createOrganizationalUnit0(newSpace))
+        .then(i => {
+          AppFormer.fireEvent(
+              new NotificationEvent({
+                type: NotificationType.SUCCESS,
+                notification: AppFormer.translate("OrganizationalUnitSaveSuccess", [
+                  AppFormer.translate("OrganizationalUnitDefaultAliasInSingular", [])
+                ])
+              })
+          );
+          return this.props.onClose();
+        })
+        .catch(() => this.setState(prevState => ({ displayedErrorMessages: prevState.errorMessages })))
+    );
+  }
+
+  private addErrorMessage(msg: string) {
+    this.setState(prevState => ({ errorMessages: [...prevState.errorMessages, msg] }));
+  }
+
+  public render() {
+    return (
+      <>
+        <Popup onClose={() => this.props.onClose()}>
```

Comment:
wouldn't this create new function each time a Popup is rendered? isn't this.props.onClose` enough? ignore if I am wrong, I still didn't manage to build this part :)

Reply:
Good question! `this.props.onClose` is not enough, because this function would've not been "bound" to this class component. They recommend binding it on the constructor, but I figured I could save this extra line by paying this little performance price. Check this source: https://reactjs.org/docs/handling-events.html

Reply:
Oh yea, you can get around this binding mechanism by defining the methods like this:

```
private addErrorMessage = (msg: string) => {
       this.setState(prevState => ({ errorMessages: [...prevState.errorMessages, msg] }));
    };
```